addons/mail/models/res_users.py

```diff
@@ -117,43 +117,54 @@ def _unsubscribe_from_channels(self):
 
     @api.model
     def systray_get_activities(self):
-        activities = self.env["mail.activity"].search([("user_id", "=", self.env.uid)])
-        activities_by_record_by_model_name = defaultdict(lambda: defaultdict(lambda: self.env["mail.activity"]))
-        for activity in activities:
-            record = self.env[activity.res_model].browse(activity.res_id)
-            activities_by_record_by_model_name[activity.res_model][record] += activity
-        model_ids = list({self.env["ir.model"]._get(name).id for name in activities_by_record_by_model_name.keys()})
+        query = """SELECT array_agg(res_id) as res_ids, m.id, count(*), act.res_model as model,
+                    CASE
+                        WHEN %(today)s::date - act.date_deadline::date = 0 Then 'today'
+                        WHEN %(today)s::date - act.date_deadline::date > 0 Then 'overdue'
+                        WHEN %(today)s::date - act.date_deadline::date < 0 Then 'planned'
+                    END AS states
+                FROM mail_activity AS act
+                JOIN ir_model AS m ON act.res_model_id = m.id
+                WHERE user_id = %(user_id)s
+                GROUP BY m.id, states, act.res_model;
```

Comment:
I don't think it should be grouped by `m.id` and `act.res_model` at the same time as both values mean the same thing?

Reply:
res_model is the name in str, it is needed in the groupby as we select it, but I'll remove it as I don't think I use it later on